### PR TITLE
Make `frozendict` import conditional.

### LINKED ISSUE #45: On Python 3.10, must use a sufficiently new frozendict version
andrewsh [noticed](https://matrix.to/#/!FeyNfPFYiGSPyHoALL:jki.re/$16371923830Lmnkh:shadura.me?via=matrix.org&via=sw1v.org&via=uhoreg.ca) the following

> FYI https://buildd.debian.org/status/fetch.php?pkg=matrix-synapse&arch=all&ver=1.47.0-1&stamp=1637184874&raw=0

```
I: pybuild base:237: python3.10 setup.py clean 
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/setup.py", line 84, in <module>
    version = exec_file(("synapse", "__init__.py"))["__version__"]
  File "/<<PKGBUILDDIR>>/setup.py", line 80, in exec_file
    exec(code, result)
  File "<string>", line 44, in <module>
  File "/usr/lib/python3/dist-packages/canonicaljson.py", line 20, in <module>
    from frozendict import frozendict
  File "/usr/lib/python3/dist-packages/frozendict/__init__.py", line 16, in <module>
    class frozendict(collections.Mapping):
AttributeError: module 'collections' has no attribute 'Mapping'
E: pybuild pybuild:354: clean: plugin distutils failed with: exit code=1: python3.10 setup.py clean 
```

The above from frozendict 1.2.

This was fine in 3.9, but failed in 3.10. The Python 3.9 docs for the collection module says:

> Deprecated since version 3.3, will be removed in version 3.10: Moved Collections Abstract Base Classes to the collections.abc module. For backwards compatibility, they continue to be visible in this module through Python 3.9.

We should support Python 3.10. I'd suggest we ought to raise the required version of frozendict. I don't know if there's a way to make that conditional on 3.10? We should start by finding out the first version of frozendict that is supported by Python 3.10; I'd guess we want the first version where `frozendict` inherits from `dict`.

(Though come to think of it, it's odd that a frozendict inherits from dict, because it has a strictly smaller feature set (no-mutability).)

### PR DESCRIPTION
**Pull request by BURG3R5**
Fixes #45

Implements the workaround suggested by @DMRobertson.

### CODE REVIEW

**Comment by BURG3R5** [comment]
If there's any other changes I need to make to complete the task, please mention.

**Comment by DMRobertson** [comment]
Many thanks for this @BURG3R5.

My thinking was a bit disconnected on #45, so to summarise:
- Python 3.10 and frozendict 1.2 are fundamentally incompatible
- It's a shame if that prevents us from using the library
- We don't actually use `frozendicts` per se---we just provide compatibility with the lib
- It'd be nice if that compatibility was optional

Could someone from the @matrix-org/synapse-core check that they agree with me? We've been bitten by thorny frozendict problems often and I'd like to make sure my proposal is sound.

If we're going to do this, we should probably mark `frozendict` as an optional dependency. That would need changes to `setup.py`; [see setuptools' documentation](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies).

**Comment by richvdh** [comment]
Yeah, I kindof agree, though also, given:
 * the `_dict` hack doesn't work with the C version of frozendict, so we have to fall back to `dict(obj)` anyway
 * nowadays most people should be using the C version of frozendict

Perhaps we should give up trying to dig into the guts of `frozendict` and just do `dict(obj)` whenever the input is a `Mapping` but not a `dict` ?

**Comment by BURG3R5** [comment]
@DMRobertson Can't we put the `._dict` task on the user? If their codebase deals with `frozendict` and is using the correct versions and implementations, then the can extricate the `._dict` themselves? Why chase that particular case?

**Comment by BURG3R5** [comment]
Also, can anyone tell me why so many CIs are failing? I don't think I changed that much...

**Comment by DMRobertson** [comment]
> Also, can anyone tell me why so many CIs are failing? I don't think I changed that much...

`coverage` is unhappy because there are new lines of source code not covered by tests. I'd try adding `# pragma: no cover` to line 25 and see if that helps.

I'd guess the a new version of `black` has released that wants to reformat existing source code. Ignore the `lint (black)` check for now; I'll try to fix it later.

**Comment by BURG3R5** [comment]
@DMRobertson That sounds good to me. One doubt: Should I make a new commit for this or revert the previous one and force push. I don't know what the preferred method is for this repo.

**Comment by DMRobertson** [comment]
> One doubt: Should I make a new commit for this or revert the previous one and force push?

Prefer new commits, please---it makes it slightly easier to review. We can always squash merge afterwards.

**Comment by BURG3R5** [comment]
@DMRobertson Can you confirm whether what I've done is what you wanted? I've turned `frozendict` imports in both the [canonicaljson.py](https://github.com/BURG3R5/python-canonicaljson/blob/frozendict_conditional/canonicaljson.py) and [test_canonicaljson.py](https://github.com/BURG3R5/python-canonicaljson/blob/frozendict_conditional/test_canonicaljson.py) files, optional. Plus I've turned the requirement mentioned in [setup.py](https://github.com/BURG3R5/python-canonicaljson/blob/27071a526eb9058c411d9aebfcc6f06b63591674/setup.py#L53) optional too.

**Comment by BURG3R5** [comment]
By the way, CI is mostly green now.

File: test_canonicaljson.py [hunk 1]
```diff
@@ -24,11 +26,15 @@
...
-
 import unittest
 from unittest import mock
 
+frozendict_type: Optional[Type]
+try:
+    from frozendict import frozendict as frozendict_type
+except ImportError:
+    frozendict_type = None
+
```
Comment (DMRobertson) [inline test_canonicaljson.py:37]: Minor: I'd be tempted to do `from canonicaljson import frozendict_type` here

File: test_canonicaljson.py [hunk 2]
```diff
@@ -109,13 +115,14 @@ def test_encode_pretty_printed(self):
...
-            encode_pretty_printed_json(frozendict({"a": 1})), b'{\n    "a": 1\n}'
-        )
+        if frozendict_type is not None:
+            self.assertEqual(
+                encode_canonical_json(frozendict_type({"a": 1})),
+                b'{"a":1}',
+            )
+            self.assertEqual(
+                encode_pretty_printed_json(frozendict_type({"a": 1})), b'{\n    "a": 1\n}'
+            )
```
Comment (DMRobertson) [inline test_canonicaljson.py:125]: I think it'd be slightly clearer to conditionally skip this test using https://docs.python.org/3/library/unittest.html#unittest.skipIf or similar.

**Review by DMRobertson — CHANGES_REQUESTED** [review]

File: test_canonicaljson.py (hunk 2 above)
Comment (BURG3R5) [reply test_canonicaljson.py]: Fixed in 7ddb3b8.

File: test_canonicaljson.py (hunk 1 above)
Comment (BURG3R5) [reply test_canonicaljson.py]: Fixed in 7ddb3b8.

**Review by DMRobertson — APPROVED** [review]
Looks good to me---many thanks!

**Pull request merged**